### ISSUE TITLE
react-native-xcode.sh doesn't filter out link local addresses correctly

### ISSUE THREAD

**Issue by msbit** [facebook/react-native#44752]
### Description

At https://github.com/facebook/react-native/blob/63e893d23d84c9bf65bad64fc359b9686eb19b4d/packages/react-native/scripts/react-native-xcode.sh#L17-L22 the script goes through a list of interfaces looking for an IP, using `ipconfig getifaddr`.

At https://github.com/facebook/react-native/blob/63e893d23d84c9bf65bad64fc359b9686eb19b4d/packages/react-native/scripts/react-native-xcode.sh#L23-L25 if the previous search hasn't found anything (`IP` is empty) it uses a looser search, `grep`ping for `inet` and discarding link local addresses (`127.x` and `169.254.x`).

The scenario I have is where my `en0` has a link local address (`169.254.36.33`) and so as the first search doesn't filter out these addresses, this is what is written to `ip.txt` at https://github.com/facebook/react-native/blob/63e893d23d84c9bf65bad64fc359b9686eb19b4d/packages/react-native/scripts/react-native-xcode.sh#L27 and so my iOS device cannot connect to the metro instance.

### Steps to reproduce

-

### React Native Version

0.74.1

### Affected Platforms

Runtime - iOS

### Output of `npx react-native info`

```text
-
```

### Stacktrace or Logs

```text
-
```

### Reproducer

-

**Comment by cortinico** [comment]
cc @cipolleschi @robhogan et al. if we need to fix this in a different way for iOS

**Comment by cipolleschi** [comment]
@msbit you have issues with your physical device only, right? It works properly on the simulator, correct?

**Comment by msbit** [comment]
@cipolleschi I haven't checked explicitly, but I'm guessing the check of platform at: https://github.com/facebook/react-native/blob/63e893d23d84c9bf65bad64fc359b9686eb19b4d/packages/react-native/scripts/react-native-xcode.sh#L16 would exclude this behaviour under simulators.

**Comment by cipolleschi** [comment]
yes, it is possible. @robhogan could you take this? I'm going to be on PTO next week. Otherwise I can look into it when I'm back.
